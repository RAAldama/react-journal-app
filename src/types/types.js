export const types = {
    login: '[Auth] Login',
    logout: '[Auth] Logout',

    uiSetError: '[Ui] SetError',
    uiRemoveError: '[Ui] RemoveError',

    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading'
}