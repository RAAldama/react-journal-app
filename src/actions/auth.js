import { types } from "../types/types";
import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { uiFinishLoading, uiStartLoading } from "./ui";

export const startLoginEmailPassword = (email, password) => {
    return ( dispatch ) => {
        dispatch( uiStartLoading() );

        firebase.auth().signInWithEmailAndPassword(email, password).then(
            ({ user }) => {
                dispatch( login(user.uid, user.displayName) );
                dispatch( uiFinishLoading() )
            }
        ).catch(
            e => {
                console.log(e);
                dispatch( uiFinishLoading() );
            }
        )
    }
}

export const startRegisterEmailPassword = (email, password, username) => {
    return ( dispatch ) => {
        firebase.auth().createUserWithEmailAndPassword(email, password).then(
            async( {user} ) => {
                await user.updateProfile({ displayName: username });

                dispatch( login(user.uid, user.displayName) );
            }
        ).catch(
            e => console.log(e)
        )
    }
}

export const startLoginGoogle = () => {
    return ( dispatch ) => {
        firebase.auth().signInWithPopup( googleAuthProvider ).then(
            ( {user} ) => {
                dispatch( login(user.uid, user.displayName) );
            }
        );
    }
}

export const login = (uid, displayName) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}