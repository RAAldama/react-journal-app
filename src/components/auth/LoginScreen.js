import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";

import { startLoginEmailPassword, startLoginGoogle } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";

export const LoginScreen = () => {

  const dispatch = useDispatch()
  const { loading } = useSelector( state => state.ui );

  const [ formValues, handleInputChange ] = useForm({
    email: '',
    password: ''
  });

  const {email, password} = formValues;

  const handleLogin = (e) => {
    e.preventDefault();

    dispatch( startLoginEmailPassword(email, password) );
  }

  const handleGoogleLogin = () => {
    dispatch( startLoginGoogle() );
  }

  return (
    <>
      <h3 className="auth__title">Login</h3>

      <form className="animate__animated animate__fadeIn" onSubmit={handleLogin}>
        <input 
          className="auth__input" 
          type="text" 
          placeholder="Correo electrónico" 
          name="email" 
          value={email} 
          onChange={handleInputChange}
          autoComplete="off" 
        />
        <input 
          className="auth__input" 
          type="password" 
          placeholder="Contraseña" 
          name="password" 
          value={password} 
          onChange={handleInputChange}
          />

        <button type="submit" className="btn btn-primary btn-block" disabled={ loading }>
          Ingresar
        </button>

        <div className="auth__social-networks">
          <p>Unete con tus redes sociales :D</p>

          <div className="google-btn" onClick={ handleGoogleLogin }>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                alt="google button"
              />
            </div>

            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link className="link" to="/auth/register">
          ¿No tienes cuenta? ¡Crea una aquí!
        </Link>
        
      </form>
    </>
  );
};
