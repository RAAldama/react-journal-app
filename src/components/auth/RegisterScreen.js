import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import validator from "validator";

import { startRegisterEmailPassword } from "../../actions/auth";
import { removeError, setError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";

export const RegisterScreen = () => {

  const dispatch = useDispatch();
  const { msgError } = useSelector( state => state.ui );

  const [formValues, handleInputChange] = useForm({
    name: 'albus',
    email: 'lel@lel.com',
    password: '123456',
    password2: '123456',
  });

  const {name, email, password, password2} = formValues;

  const handleRegister = (e) => {
    e.preventDefault();

    if( isFormValid() ){
      dispatch( startRegisterEmailPassword(email, password, name) );
    }
  }

  const isFormValid = () => {
    if( name.trim().length === 0 ){
      const error = 'El NOMBRE es obligatorio';
      dispatch( setError(error) );

      return false;

    } else if( !validator.isEmail(email) ) {
      const error = 'El CORREO debe de ser válido';
      dispatch( setError(error) );

      return false;

    } else if( password !== password2 || password.length < 5 ){
      const error = 'Revise la CONTRASEÑA';
      dispatch( setError(error) );

      return false;
    }

    dispatch( removeError() );

    return true;
  }

  return (
    <>
      <h3 className="auth__title">Register</h3>

      <form className="animate__animated animate__fadeIn" onSubmit={handleRegister}>
        
        {
          msgError &&
          <div className="auth__alert-error">
            { msgError }
          </div>
        }

        <input
          className="auth__input"
          type="text"
          placeholder="Usuario"
          name="name"
          value={name} 
          onChange={handleInputChange}
          autoComplete="off"
        />
        <input
          className="auth__input"
          type="text"
          placeholder="Correo electrónico"
          name="email"
          value={email} 
          onChange={handleInputChange}
          autoComplete="off"
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Contraseña"
          name="password"
          value={password} 
          onChange={handleInputChange}
        />
        <input
          className="auth__input"
          type="password"
          placeholder="Confirmar contraseña"
          name="password2"
          value={password2} 
          onChange={handleInputChange}
        />

        <button type="submit" className="btn btn-primary btn-block mb-5">
          Registrar
        </button>

        <Link className="link" to="/auth/login">
          ¿Con cuenta? ¡Ingresa aquí!
        </Link>
      </form>
    </>
  );
};
