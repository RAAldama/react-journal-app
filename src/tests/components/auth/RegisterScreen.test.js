import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store'
import thunk from "redux-thunk"
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
}

let store = mockStore(initState);

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>

            <RegisterScreen />

        </MemoryRouter>
    </Provider>
)

describe('tests on <RegisterScreen />', () => {
    
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })
    
    test('should make the respective dispatch of action', () => {
        const emailField = wrapper.find('input[name="email"]');
        emailField.simulate('change', {
            target: {
                value: '',
                name:'email'
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();
        expect( actions[0] ).toEqual({ type: types.uiSetError, payload: 'El CORREO debe de ser válido' });
    })

    test('should render the danger box with error', () => {
        
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'El CORREO debe de ser válido'
            }
        }
        
        const store = mockStore(initState);
        
        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
        
                    <RegisterScreen />
        
                </MemoryRouter>
            </Provider>
        )

        expect( wrapper.find('.auth__alert-error').exists() ).toBe(true);
        expect( wrapper.find('.auth__alert-error').text().trim() ).toBe( initState.ui.msgError );

    })
    
})
