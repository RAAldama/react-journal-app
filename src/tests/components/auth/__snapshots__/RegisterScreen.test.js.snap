// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <RegisterScreen /> should render correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Usuario"
      type="text"
      value="albus"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Correo electrónico"
      type="text"
      value="lel@lel.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Contraseña"
      type="password"
      value="123456"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirmar contraseña"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Registrar
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      ¿Con cuenta? ¡Ingresa aquí!
    </a>
  </form>,
]
`;
